### PR TITLE
Fix DistributionForecast failure on GPU

### PR DESCRIPTION
**Pull request by huibinshen**
*Issue #, if available:*

*Description of changes:*

Fixing https://github.com/awslabs/gluonts/issues/2662
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Please tag this pr with at least one of these labels to make our release process faster:** BREAKING, new feature, bug fix, other change, dev setup

### CODE REVIEW

File: src/gluonts/evaluation/_base.py [hunk 1]
```diff
@@ -256,6 +257,10 @@ def __call__(
             desc="Running evaluation",
         ) as it, np.errstate(divide="ignore", invalid="ignore"):
             if self.num_workers and not sys.platform == "win32":
+
+                if torch.cuda.is_available():
+                    torch.multiprocessing.set_start_method('spawn', force=True)
+
```
Comment (lostella) [inline src/gluonts/evaluation/_base.py]: I think this should not happen here, it's probably better to keep it on the caller side

File: test/test_on_gpu/test_torch_model.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+
+@pytest.mark.parametrize(
+    "torch_model_cls",
+    [
+        SimpleFeedForwardEstimator,
+        DeepAREstimator
+    ]
+)
+def test_torch_model_on_gpu(
```
Comment (lostella) [inline test/test_on_gpu/test_torch_model.py]: You can condition the execution of this test on `torch.cuda.is_available()` using `skipif` from `pytest`: https://docs.pytest.org/en/7.1.x/how-to/skipping.html#id1

File: test/test_on_gpu/test_torch_model.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    torch_model_cls: torch.model.PyTorchLightningEstimator
```
Comment (lostella) [inline test/test_on_gpu/test_torch_model.py]: You're passing a class now (so, a `Type` and not a `PyTorchLightningEstimator`), but you could indeed pass the estimator object directly

File: src/gluonts/evaluation/_base.py (hunk 1 above)
Comment (huibinshen) [reply src/gluonts/evaluation/_base.py]: For now I put it into `src/gluonts/torch/__init__.py`, what do you think of it? Or is it better to put it as part of the test you mean?

File: test/test_on_gpu/test_torch_model.py (hunk 3 above)
Comment (huibinshen) [reply test/test_on_gpu/test_torch_model.py]: What is the preferred way in this case?

File: src/gluonts/evaluation/_base.py (hunk 1 above)
Comment (lostella) [reply src/gluonts/evaluation/_base.py]: Yes, maybe part of the test is better

File: test/test_on_gpu/test_torch_model.py (hunk 3 above)
Comment (lostella) [reply test/test_on_gpu/test_torch_model.py]: I think passing the estiamator object (so, creating it as part of the `pytest.mark.parametrize` decorator)

File: src/gluonts/torch/__init__.py [hunk 4]
```diff
@@ -19,7 +19,10 @@
     "SimpleFeedForwardEstimator",
     "TemporalFusionTransformerEstimator",
 ]
+import torch
 
+if torch.cuda.is_available():
+    torch.multiprocessing.set_start_method("spawn", force=True)
```
Comment (jaheba) [inline src/gluonts/torch/__init__.py]: You generally want to do this after all imports, not mix them in.

File: src/gluonts/evaluation/_base.py [hunk 5]
```diff
@@ -256,6 +256,7 @@ def __call__(
             desc="Running evaluation",
         ) as it, np.errstate(divide="ignore", invalid="ignore"):
             if self.num_workers and not sys.platform == "win32":
+
```
Comment (jaheba) [inline src/gluonts/evaluation/_base.py]: Let's revert this change :)

File: src/gluonts/torch/model/forecast.py [hunk 6]
```diff
@@ -86,7 +86,12 @@ def mean_ts(self) -> pd.Series:
 
     def quantile(self, level: Union[float, str]) -> np.ndarray:
         level = Quantile.parse(level).value
-        return self.distribution.icdf(torch.tensor([level])).cpu().numpy()
+        device = self.distribution.mean.device
+        return (
+            self.distribution.icdf(torch.tensor([level], device=device))
```
Comment (jaheba) [inline src/gluonts/torch/model/forecast.py]: I think I would just inline `self.distribution.mean.device` here.

File: test/test_on_gpu/test_torch_model.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+from gluonts.dataset.pandas import PandasDataset
+from gluonts.evaluation import make_evaluation_predictions
+from gluonts.torch.model.estimator import PyTorchLightningEstimator
+from gluonts.torch.model.simple_feedforward import SimpleFeedForwardEstimator
+from gluonts.torch.model.deepar import DeepAREstimator
+from gluonts.evaluation import Evaluator
+
+
+@pytest.mark.skipif(
+    torch.cuda.is_available() is False, reason="only run the test on GPU"
```
Comment (jaheba) [inline test/test_on_gpu/test_torch_model.py]:
```suggestion
    not torch.cuda.is_available(), reason="only run the test on GPU"
```

File: src/gluonts/evaluation/_base.py (hunk 1 above)
Comment (huibinshen) [reply src/gluonts/evaluation/_base.py]: If I leave it to the test function, the users may still face the error when using torch models.

File: src/gluonts/torch/__init__.py (hunk 4 above)
Comment (huibinshen) [reply src/gluonts/torch/__init__.py]: Do you mean putting it at the end of the `torch/__init__.py` file?

File: src/gluonts/evaluation/_base.py (hunk 5 above)
Comment (huibinshen) [reply src/gluonts/evaluation/_base.py]: That would leave the following fix not necessary
```
if torch.cuda.is_available():
    torch.multiprocessing.set_start_method("spawn", force=True)
```
Are we sure that we want to do this? What's your input? @lostella

Comment (jaheba) [reply src/gluonts/evaluation/_base.py]: No, I mean the empty line that you've introduced ;)

File: src/gluonts/torch/__init__.py (hunk 4 above)
Comment (jaheba) [reply src/gluonts/torch/__init__.py]: Somewhere below the imports.

File: src/gluonts/torch/model/forecast.py [hunk 8]
```diff
@@ -86,7 +86,13 @@ def mean_ts(self) -> pd.Series:
 
     def quantile(self, level: Union[float, str]) -> np.ndarray:
         level = Quantile.parse(level).value
-        return self.distribution.icdf(torch.tensor([level])).cpu().numpy()
+        return (
+            self.distribution.icdf(
+                torch.tensor([level], device=self.distribution.mean.device)
```
Comment (jaheba) [inline src/gluonts/torch/model/forecast.py:91]: Something we could also do is to add `.device` to the distribution class, that would this a bit cleaner in my view.

Comment (huibinshen) [reply src/gluonts/torch/model/forecast.py:91]: `Distribution` is from torch, I could add a `.device` into the `AffineTransformed` class.

**Comment by jaheba** [comment]
Looks generally good to me, did you try running this on GPU and it succeeds?

**Comment by jaheba** [comment]
I think my suggestion with the `device` things was actually a bad one, sorry!

**Comment by huibinshen** [comment]
> Looks generally good to me, did you try running this on GPU and it succeeds?

I ran the test on a GPU instance and it was fine.

**Review by lostella — APPROVED** [review]
